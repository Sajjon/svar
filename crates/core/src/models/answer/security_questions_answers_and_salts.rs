use crate::prelude::*;

#[derive(
    Clone,
    PartialEq,
    Eq,
    derive_more::Debug,
    derive_more::Display,
    derive_more::Deref,
    derive_more::DerefMut,
    derive_more::From,
)]
#[display("SecurityQuestionsAnswersAndSalts({})", self.0.len())]
pub struct SecurityQuestionsAnswersAndSalts<const QUESTION_COUNT: usize>(
    [SecurityQuestionAnswerAndSalt; QUESTION_COUNT],
);

impl<const QUESTION_COUNT: usize>
    SecurityQuestionsAnswersAndSalts<QUESTION_COUNT>
{
    pub fn try_from_iter(
        qas: impl IntoIterator<Item = SecurityQuestionAnswerAndSalt>,
    ) -> Result<Self> {
        let qas = qas.into_iter().collect::<IndexSet<_>>();
        let len = qas.len();
        let arr: [SecurityQuestionAnswerAndSalt; QUESTION_COUNT] = qas
            .into_iter()
            .collect::<Vec<_>>()
            .try_into()
            .map_err(|_| Error::InvalidQuestionsAndAnswersCount {
                expected: QUESTION_COUNT,
                found: len,
            })?;

        Ok(Self(arr))
    }
}

#[cfg(test)]
impl SecurityQuestionsAnswersAndSalts<6> {
    pub(crate) fn sample_wrong_answers() -> Self {
        type Q = SecurityQuestion;
        type QA = SecurityQuestionAnswerAndSalt;
        Self::try_from_iter([
            QA {
                question: Q::failed_exam(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_aced(),
            },
            QA {
                question: Q::parents_met(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_babe(),
            },
            QA {
                question: Q::first_concert(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_cafe(),
            },
            QA {
                question: Q::first_kiss_whom(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_dead(),
            },
            QA {
                question: Q::first_kiss_location(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_ecad(),
            },
            QA {
                question: Q::spouse_met(),
                answer: "Wrong answer".to_owned(),
                salt: Exactly32Bytes::sample_fade(),
            },
        ])
        .expect("Should have been 6 questions and answers")
    }
}

impl HasSampleValues for SecurityQuestionsAnswersAndSalts<6> {
    fn sample() -> Self {
        type Q = SecurityQuestion;
        type QA = SecurityQuestionAnswerAndSalt;
        Self::try_from_iter([
            QA {
                question: Q::failed_exam(),
                answer: "MIT, year 4, Python".to_owned(),
                salt: Exactly32Bytes::sample_aced(),
            },
            QA {
                question: Q::parents_met(),
                answer: "London, 1973".to_owned(),
                salt: Exactly32Bytes::sample_babe(),
            },
            QA {
                question: Q::first_concert(),
                answer: "Jean-Michel Jarre, Paris La DÃ©fense, 1990".to_owned(),
                salt: Exactly32Bytes::sample_cafe(),
            },
            QA {
                question: Q::first_kiss_whom(),
                answer: "John Doe".to_owned(),
                salt: Exactly32Bytes::sample_dead(),
            },
            QA {
                question: Q::first_kiss_location(),
                answer: "Behind the shed in the oak tree forrest.".to_owned(),
                salt: Exactly32Bytes::sample_ecad(),
            },
            QA {
                question: Q::spouse_met(),
                answer: "Tokyo, 1989".to_owned(),
                salt: Exactly32Bytes::sample_fade(),
            },
        ])
        .expect("Should have been 6 questions and answers")
    }

    fn sample_other() -> Self {
        type Q = SecurityQuestion;
        type QA = SecurityQuestionAnswerAndSalt;
        Self::try_from_iter([
            QA {
                question: Q::child_middle_name(),
                answer: "Joe".to_owned(),
                salt: Exactly32Bytes::sample_aced(),
            },
            QA {
                question: Q::stuffed_animal(),
                answer: "Bobby".to_owned(),
                salt: Exactly32Bytes::sample_babe(),
            },
            QA {
                question: Q::oldest_cousin(),
                answer: "Roxanne".to_owned(),
                salt: Exactly32Bytes::sample_cafe(),
            },
            QA {
                question: Q::teacher_grade3(),
                answer: "Ali".to_owned(),
                salt: Exactly32Bytes::sample_dead(),
            },
            QA {
                question: Q::applied_uni_no_attend(),
                answer: "Oxford".to_owned(),
                salt: Exactly32Bytes::sample_ecad(),
            },
            QA {
                question: Q::first_school(),
                answer: "Hogwartz".to_owned(),
                salt: Exactly32Bytes::sample_fade(),
            },
        ])
        .expect("Should have been 6 questions and answers")
    }
}
