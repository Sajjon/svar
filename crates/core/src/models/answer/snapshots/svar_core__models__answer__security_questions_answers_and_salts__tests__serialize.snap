---
source: crates/core/src/models/answer/security_questions_answers_and_salts.rs
expression: "Sut::sample()"
---
[
  {
    "question": {
      "id": 0,
      "version": 1,
      "kind": "Freeform",
      "question": "What was the first exam you failed",
      "expected_answer_format": {
        "answer_structure": "<SCHOOL>, <SCHOOL_GRADE>, <SUBJECT>",
        "example_answer": "MIT, year 4, Python",
        "unsafe_answers": []
      }
    },
    "answer": "MIT, year 4, Python",
    "salt": "acedacedacedacedacedacedacedacedacedacedacedacedacedacedacedaced"
  },
  {
    "question": {
      "id": 1,
      "version": 1,
      "kind": "Freeform",
      "question": "In which city and which year did your parents meet?",
      "expected_answer_format": {
        "answer_structure": "<CITY>, <YEAR>",
        "example_answer": "Berlin, 1976",
        "unsafe_answers": []
      }
    },
    "answer": "London, 1973",
    "salt": "babebabebabebabebabebabebabebabebabebabebabebabebabebabebabebabe"
  },
  {
    "question": {
      "id": 2,
      "version": 1,
      "kind": "Freeform",
      "question": "What was the first concert you attended?",
      "expected_answer_format": {
        "answer_structure": "<ARTIST>, <LOCATION>, <YEAR>",
        "example_answer": "Jean-Michel Jarre, Paris La Défense, 1990",
        "unsafe_answers": []
      }
    },
    "answer": "Jean-Michel Jarre, Paris La Défense, 1990",
    "salt": "cafecafecafecafecafecafecafecafecafecafecafecafecafecafecafecafe"
  },
  {
    "question": {
      "id": 3,
      "version": 1,
      "kind": "Freeform",
      "question": "What was the name of the boy or the girl you first kissed?",
      "expected_answer_format": {
        "answer_structure": "<NAME>",
        "example_answer": "Maria",
        "unsafe_answers": []
      }
    },
    "answer": "John Doe",
    "salt": "deaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddeaddead"
  },
  {
    "question": {
      "id": 4,
      "version": 1,
      "kind": "Freeform",
      "question": "Where were you when you had your first kiss?",
      "expected_answer_format": {
        "answer_structure": "<LOCATION>",
        "example_answer": "At bus stop outside of Dallas",
        "unsafe_answers": [
          "Specifying only a country as location would be unsafe"
        ]
      }
    },
    "answer": "Behind the shed in the oak tree forrest.",
    "salt": "ecadecadecadecadecadecadecadecadecadecadecadecadecadecadecadecad"
  },
  {
    "question": {
      "id": 5,
      "version": 1,
      "kind": "Freeform",
      "question": "In what city and which year did you meet your spouse/significant other?",
      "expected_answer_format": {
        "answer_structure": "<CITY>, <YEAR>",
        "example_answer": "Berlin, 1976",
        "unsafe_answers": []
      }
    },
    "answer": "Tokyo, 1989",
    "salt": "fadefadefadefadefadefadefadefadefadefadefadefadefadefadefadefade"
  }
]
